docs/latest/design/egctl.md

```diff
@@ -0,0 +1,73 @@
+# Introduce egctl
+
+## Motivation
+
+EG should provide a command line tool with following capabilities:
+
+- One-step installation/uninstallation experience
+- Collect configuration from envoy proxy and gateway
+- Update admin options such as changing log level
+- Analyse system configuration to diagnose any issues in envoy gateway
+
+This tool is named `egctl`.
+
+## Syntax
+
+Use the following syntax to run `egctl` commands from your terminal window:
+
+```console
+egctl [command] [entity] [name] [flags]
+```
+
+where `command`, `name`, and `flags` are:
+
+* `command`: Specifies the operation that you want to perform on one or more resources,
+  for example `install`, `uninstall`, `config`, `version`.
+
+* `entity`: Specifies the entity the operation is being performed on such as `envoy-proxy` or `envoy-gateway`.
+
+* `name`: Specifies the name of the specified instance. 
+
+* `flags`: Specifies optional flags. For example, you can use the `-c` or `--config` flags to specify the values for installing.
+
+If you need help, run `kubectl help` from the terminal window.
+
+## Operation
+
+The following table includes short descriptions and the general syntax for all the `egctl` operations:
+
+| Operation   | Syntax                           | Description                                                                 |
+| ----------- | -------------------------------- | --------------------------------------------------------------------------- |
+| `version`   | `egctl version`                  | Prints out build version information.                                       |
+| `install`   | `egctl install -c CUSTOMPROFILE` | Install or reconfigure EG on a cluster.                                     |
+| `uninstall` | `egctl uninstall`                | Uninstall EG from a cluster                                                 |
+| `config`    | `egctl config ENTITY`            | Retrieve information about proxy configuration from envoy proxy and gateway |
+| `log`       | `egctl log ENTITY --level trace` | Change envoy proxy's log level                                              |
```

Comment:
I would expect `log` to also support changing EG's log level. I also prefer `logs` over `log`. Can `log` be used to tail the logs of EG or an Envoy Proxy instance? For example:
```
egctl log envoy-proxy <name> -f
```

Reply:
`log` only change EG's log level, did `envoy-gateway` support change log level on the fly?

BTW, I donot think we should support `-f`, you can use `kubectl logs` in kubernetes world, we may extend when other provider is supported, but not now.

Reply:
> log only change EG's log level, did envoy-gateway support change log level on the fly?

Changing the EG log level will cause a rolling restart.

> BTW, I donot think we should support -f, you can use kubectl logs in kubernetes world, we may extend when other provider is supported, but not now.

I'm fine using kubectl until EG supports other providers.

I have given additional thought to supporting the `log` subcommand. IMHO we should drive all config changes through the EnvoyProxy API. Otherwise, we set a precedent for supporting additional config knobs in the future which increases the support burden. Additionally, having one way to manage config provides a simple mental model for users.

If a user would like to change the log level, we should support:
```
cat <<EOF | egctl apply -f -
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: example-config
spec:
  logging:
    ...
EOF
```

Reply:
A rolling restart with control plane(envoy-gateway) is ok in mostly usecases, but it's a little danger on data plane(envoy-proxy), especially TCP.

After you change the config CRD, you can apply changes with `install`.

Reply:
I agree that a rolling restart of EG is a non-issue. However, changing Envoy's log level will cause a restart since the Envoy deployment `--log-level` CLI argument will be changed. Can Envoy avoid the restart if the log-level configuration was accomplished without using the CLI argument?

Reply:
sure, see https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--logging